noise-research/secure-transfer.md

```diff
@@ -0,0 +1,235 @@
+
+The requests is made by exposing a QR code that, by default, has to be scanned by device `A`. 
+If device `A` doesn't have a camera while device `B` does, [it is possible](#Rationale) to execute a slightly different pairing (with same security guarantees), where `A` is exposing a QR code instead.
+
+### Employed Cryptographic Primitives
+
+- `H`: the underlying hash function, i.e. SHA-256;
+- `HKDF`: a key derivation function (based on SHA-256);
+- `Curve25519`: the underlying elliptic curve for Diffie-Hellman (DH) operations.
+
+### The `WakuPairing` Noise Handshake
+The devices execute a custom handshake derived from `X1X1`, where they mutually exchange and authenticate their device static keys, i.e.
+
+```
+WakuPairing:
+0.   <- eB              {H(sB||r), contentTopic}
+     ...
+1.   -> eA, eAeB        {H(sA||s)}   [auth_code]
+2.   <- sB, eAsB        {r}
+3.   -> sA, sAeB, sAsB  {s}
+
+{}: payload,    []: user interaction
+```
+
+
+Beside the ephemeral key, all the information embedded in the QR code should be passed to the prologue of the Noise handshake (e.g. `H(sB||r)`, `contentTopic`, etc.).
+
+
+### Protocol Flow
+1. The device `B` exposes through a QR code a Base64 serialization of:
+    - An ephemeral public key `eB`;
+    - A `contentTopic` where the information exchange will take place. `contentTopic` follows [23/WAKU2-TOPICS](https://rfc.vac.dev/spec/23/#content-topics) specifications and is then of the form `/{application-name}/{application-version}/wakunoise/1/sessions-{shard-id}/{random-message-id}/proto` for a randomly generated `random-message-id`. `contentTopic` can be then serialized in compressed form as `{application-name}:{application-version}:{shard-id}:{random-message-id}`.
+    - A commitment `H(sB||r)` for its static key `sB` where `r` is a random fixed-lenght value.
+
+2. The device `A`:
+    - scans the QR code;
+    - obtains `eB`, `contentTopic`, `Hash(sB|r)`;
+    - initializes the Noise handshake by passing `contentTopic` and `Hash(sB||r)` to the handshake prologue;
+    - executes the pre-handshake message, i.e. processes the key `eB`;
+    - executes the first handshake message over `contentTopic`, i.e. 
+        - processes and sends a Waku message containing an ephemeral key `eA`; 
+        - performs `DH(eA,eB)` (which computes a symmetric encryption key);
+        - attach as payload to the handshake message a commitment `H(sA|s)` for `A`'s static key `sA`, where `s` is a random fixed-lenght value;
+    - an 8-digits authorization code `auth_code` obtained as `HKDF(h) mod 10^8` is displayed on the device, where `h`is the handshake value obtained once the first handshake message is processed.
+
+3. The device `B`:
+    - subscribes to any message sent to `/{application-name}/{application-version}/wakunoise/1/sessions-{shard-id}/*` and locally filters only messages sent to `contentTopic`. If any, continues.
```

Comment:
> Another thing is that `filter` protocol does not yet support subscribing to partial content topic matches like `/{application-name}/{application-version}/wakunoise/1/sessions-{shard-id}/*`, but that shouldn't be a blocker for secure transfer.

Good point, AFAIK, store does not allow retrieval either of partial/globbing topics.

Same applies to any js/nim/go API to process inbound messages: a content topic `string` is expected, not a glob or regex.

Why do we need a message id in the content topic? I'd avoid it if possible.

Reply:
> If this is to become a spec (which I think it should :) ), wondering if it's not better to avoid prescribing things like "subscribe" and "filter".

I agree.I would suggest using _listen to `contentTopic`_ for example.